R/namespace_linter.R

```diff
@@ -33,97 +33,124 @@ namespace_linter <- function(check_exports = TRUE, check_nonexports = TRUE) {
-                    column_number = col1,
-                    type = "style",
-                    message = sprintf("'%s' is exported from {%s}. Use %s::%s instead.",
-                      syms[[i]], pkgs[[i]], pkgs[[i]], syms[[i]]),
-                    line = source_expression$file_lines[[line1]],
-                    ranges = list(c(col1, col2))
-                  ))
-                }
-              } else {
-                line1 <- as.integer(xml2::xml_attr(sym_nodes[[i]], "line1"))
-                col1 <- as.integer(xml2::xml_attr(sym_nodes[[i]], "col1"))
-                col2 <- as.integer(xml2::xml_attr(sym_nodes[[i]], "col2"))
-                return(Lint(
-                  filename = source_expression$filename,
-                  line_number = line1,
-                  column_number = col1,
-                  type = "warning",
-                  message = sprintf("'%s' does not exist in {%s}.", syms[[i]], pkgs[[i]]),
-                  line = source_expression$file_lines[[line1]],
-                  ranges = list(c(col1, col2))
-                ))
-              }
-            }
-          } else {
-            # nocov start: need getNamespace to fail. Could be done with {mockery} in principle
-            line1 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "line1"))
-            col1 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "col1"))
-            col2 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "col2"))
-            return(Lint(
-              filename = source_expression$filename,
-              line_number = line1,
-              column_number = col1,
-              type = "warning",
-              message = conditionMessage(ns),
-              line = source_expression$file_lines[[line1]],
-              ranges = list(c(col1, col2))
-            ))
-            # nocov end
-          }
-        }
-      } else {
-        line1 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "line1"))
-        col1 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "col1"))
-        col2 <- as.integer(xml2::xml_attr(pkg_nodes[[i]], "col2"))
-        return(Lint(
-          filename = source_expression$filename,
-          line_number = line1,
-          column_number = col1,
-          type = "warning",
-          message = sprintf("Package '%s' is not installed.", pkgs[[i]]),
-          line = source_expression$file_lines[[line1]],
-          ranges = list(c(col1, col2))
-        ))
+    results <- lapply(
+      seq_along(ns_nodes),
+      function(i) {
+        build_namespace_lint(pkgs[[i]], syms[[i]], pkg_nodes[[i]], sym_nodes[[i]], check_exports, check_nonexports)
       }
-    })
+    )
 
     results[!vapply(results, is.null, logical(1L))]
   })
 }
+
+build_namespace_lint <- function(package, symbol, package_node, symbol_node, check_exports, check_nonexports) {
+  if (!package %in% installed_packages) {
+    return(build_uninstalled_lint(package, package_node, source_expression))
+  }
+  if (!check_exports && !check_nonexports) {
```

Comment:
Not sure if `cyclocomp()` knows that `return()` is terminal. Maybe it'll report a better value if these are `else if`s?

Reply:
(completely refactored again)